### ISSUE TITLE
[FEATURE] Docker-Image als Multi-Plattform anbieten

### ISSUE THREAD

**Issue by cagix** [Compilerbau/Lecture#110]
Das aktuelle Docker-Image läuft nur auf Intel-Maschinen. Es wäre gut, wenn es auch auf der Mac M1-Plattform laufen würde. Was muss an der Definition des Images geändert werden?

**Comment by cagix** [comment]
@jposselt @Fabian-Hu Könnt ihr das bitte recherchieren und Lösungsvorschläge vorbereiten? Ausprobieren geht mangels M1 ja sicher nicht, oder? Ich glaube, M1 fällt unter "ARM64" oder so.

**Comment by jposselt** [comment]
Man müsste wahrscheinlich auf einem Docker-Image für die entsprechende Architektur aufbauen. Unser momentanes Basis-Image ([pandoc/latex](https://hub.docker.com/r/pandoc/latex)) gibt es aber nur für amd64. Vielleicht könnte man sowas wie [minidocks/pandoc](https://hub.docker.com/r/minidocks/pandoc) verwenden. Das hat Varianten für amd64 und arm64.

**Comment by cagix** [comment]
muss man das explizit formulieren oder (wenn man das richtige basis-image auswählt) geht das dann beim builden automatisch? und gibts von pandoc/docker nicht auch arm-images?

**Comment by jposselt** [comment]
> 
> 
> muss man das explizit formulieren oder (wenn man das richtige basis-image auswählt) geht das dann beim builden automatisch? 

Das weiß ich noch nicht.

>
> und gibts von pandoc/docker nicht auch arm-images?

Ich habe jetzt auf die Schnelle keines gefunden.

**Comment by jposselt** [comment]
@cagix Hast du einen Computer mit ARM Architektur?. Wenn ja könntest du mal folgendes ausprobieren:

1. `docker pull minidocks/pandoc`
2. `docker image inspect minidocks/pandoc -f '{{.Architecture}}'`

**Comment by cagix** [comment]
> @cagix Hast du einen Computer mit ARM Architektur?. Wenn ja könntest du mal folgendes ausprobieren:
> 
> 1. `docker pull minidocks/pandoc`
> 2. `docker image inspect minidocks/pandoc -f '{{.Architecture}}'`

ich leider (glücklicherweise?) nicht, aber @cyildiz hat einen ... :)

@cyildiz: kannst du das mal bei gelegenheit ausprobieren?

**Comment by cyildiz** [comment]
> @cyildiz: kannst du das mal bei gelegenheit ausprobieren?

Ausgabe: 
`arm64`

**Comment by cagix** [comment]
cool, das scheint zu gehen :)

@jposselt dann könnte man das als basis nehmen und müsste dann latex etc. manuell nachinstallieren. bzw. für @cyildiz würde eigentlich die kombination pandoc und hugo reichen, die erzeugung der foliensätze braucht sie ja erstmal nicht.

**Comment by cyildiz** [comment]
Ich habe während der Docker-Installation auf die Anforderung des Systems hin [Rosetta](https://www.macrumors.com/how-to/install-rosetta-m1-apple-silicon-mac/) installiert. Das scheint aber bei dem make nicht zu helfen. Oder ist das was ganz anderes?

**Comment by cagix** [comment]
> Ich habe während der Docker-Installation auf die Anforderung des Systems hin [Rosetta](https://www.macrumors.com/how-to/install-rosetta-m1-apple-silicon-mac/) installiert. Das scheint aber bei dem make nicht zu helfen. Oder ist das was ganz anderes?

hmmm, idealerweise läuft docker und das image nativ als arm64. wenn ich das richtig verstanden habe, müsste sonst docker in rosetta laufen, und dann könntest du die "normalen" intel-images nehmen. aber das dürfte dann extrem langsam sein, weil ja alles erst noch übersetzt werden muss.

make selbst sollte davon unabhängig sein, das ruft ja nur die docker-container auf.

**Comment by cyildiz** [comment]
Es gibt [Docker für Apple silicon](https://docs.docker.com/desktop/mac/apple-silicon/), aber der braucht anscheinend auch Rosetta, weil:
> some binaries are still Darwin/AMD64

Außerdem:
> Not all images are available for ARM64 architecture. You can add `--platform linux/amd64` to run an Intel image under emulation. 

> However, attempts to run Intel-based containers on Apple Silicon machines under emulation can crash as qemu sometimes fails to run the container. ... Even when the containers do run correctly under emulation, they will be slower and use more memory than the native equivalent.

> We are encouraging container authors to produce arm64, or multi-arch, versions of their containers.

Ist hier [so was ähnliches](https://jitsu.com/blog/multi-platform-docker-builds) gemeint?

**Comment by cagix** [comment]
ja, ich denke schon. leider wird unser basis-image nur für intel angeboten, aber das könnte man reparieren, indem man ein arm-image nimmt und dort pandoc und hugo reintut (jeweils in der arm-variante). 

alternativ ziehe dir einfach pandoc und hugo für arm und installiere die direkt. ich glaube, in ki haben wir keine tex-abhängigkeiten außer meinen folien (und die musst du ja nicht erzeugen).

**Comment by cyildiz** [comment]
> alternativ ziehe dir einfach pandoc und hugo für arm und installiere die direkt.

wenn ich diese lokal installiere und kein docker verwende, dann muss ich bei make etwas anders machen, oder?
Du hattest etwas ähnliches erwähnt, ich konnte aber die email nicht finden.

**Comment by cagix** [comment]
> > alternativ ziehe dir einfach pandoc und hugo für arm und installiere die direkt.
> 
> wenn ich diese lokal installiere und kein docker verwende, dann muss ich bei make etwas anders machen, oder? Du hattest etwas ähnliches erwähnt, ich konnte aber die email nicht finden.

steht im Makefile oben als Kommentar drin: Du musst eine Umgebungsvariable "DOCKER" mit dem Wert "false" in der Shell setzen, oder das beim Make-Aufruf mit erledigen: "export DOCKER=false; make TARGET"

**Comment by cyildiz** [comment]
Vielen Dank! Das hat super geklappt.
